Tracing: Migrate to OpenTelemetry library

matej-g opened the pull request:
<!--
    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --sign-off flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - No tests are needed for internal implementation changes.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

Since the Jaeger Go client has now [officially been sunset](https://github.com/jaegertracing/jaeger-client-go#-this-library-is-being-deprecated) and OpenTelemetry's tracing now has a [stable release](https://opentelemetry.io/docs/go/), the tracer implementation and tracing instrumentation should be moved to OpenTelemetry and switched to exporting to OTLP (OpenTelemetry Collector) instead of Jager. Since I was instructed that the tracing was never officially supported and that there is no need for backwards compatibility, this change:
- Provides new `enable-feature` flag and configuration for setting up tracing
- Replaces the OpenTracing tracer by a tracing manager, which takes care of creating, (re)installing and shutting down the Otel tracer provider
- Enables tracing provider 'reinstallation' with configuration reload
- Replaces instrumentation with corresponding equivalent constructs from the OpenTelemtry library
- Replaces the HTTP client and server instrumentation with corresponding OpenTelemetry library ([`otelhttp`](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation/net/http/otelhttp))

Following merging this change, update of the documentation to include tracing configuration would follow as well.

Resolves #9644.

roidelapluie commented:
I am completely lost here.

I was thinking that OpenTelemetry means that we do not need to have jaeger instrumentation, but that it would be available to all otel compatible backands. What did I miss?

roidelapluie commented:
Thanks for your contribution!

Can we leave jaeger out of this and only provide OTLP support? Users can use the OpenTelemetry collector to push to the tracing backend they want then.

It would be great to have this (reloadable) in configuration file and enabled with a feature flag (see --enable-feature).

Thanks!

matej-g commented:
> Can we leave jaeger out of this and only provide OTLP support? Users can use the OpenTelemetry collector to push to the tracing backend they want then.
> 
> It would be great to have this (reloadable) in configuration file and enabled with a feature flag (see --enable-feature).

Hey @roidelapluie, thanks for taking a look! I see the original comment disappeared so I think now we are on the same page - we want to export to OTLP instead and not to Jaeger, as was the case until now (My initial impression was we just want to switch libraries, but functionally keep exporting to same backend - Jaeger). From my side this makes totally sense and I'm happy to adjust the flags / reloading as well. I'll be back with more commits :+1:

matej-g commented:
@roidelapluie I reworked this a bit, now exporting to OTLP and configurable via `enable-feature` flag and config file. Feedback is welcome!

roidelapluie commented:
Thanks. There are two major points to start with that I have identified:

we should error if the user specifies a tracing configuration and the flag is not enabled.

The documentation is missing (should be in the feature flags page).

I'll deep review after this, thanks.

matej-g commented:
> Thanks. There are two major points to start with that I have identified:
> 
> we should error if the user specifies a tracing configuration and the flag is not enabled.
> 
> The documentation is missing (should be in the feature flags page).

Understood, I wanted to separate documentation from the PR with code as per description, but I'd be happy to add it here if that's fine for reviewers to review both in one go.

LeviHarrison commented on config/config.go:
```diff
@@ -499,6 +500,15 @@ type StorageConfig struct {
 
+// TracingConfig configures the tracing options.
+type TracingConfig struct {
+	GrpcClient       bool    `yaml:"grpc_client,omitempty"`
```
We should preserve the capitalization of "[g]RPC".

```suggestion
	GRPCClient       bool    `yaml:"grpc_client,omitempty"`
```

LeviHarrison commented on cmd/prometheus/main.go:
```diff
@@ -193,6 +191,9 @@ func (c *flagConfig) setFeatureListOptions(logger log.Logger) error {
 				level.Info(logger).Log("msg", "Experimental agent mode enabled.")
+			case "tracing":
+				c.enableTracing = true
+				level.Info(logger).Log("msg", "Tracinge enabled.")
```
```suggestion
				level.Info(logger).Log("msg", "Tracing enabled.")
```

LeviHarrison commented on tracing/tracing.go:
```diff
@@ -0,0 +1,171 @@
+	m.config = cfg.TracingConfig
+	otel.SetTracerProvider(tp)
+
+	level.Info(m.logger).Log("msg", "Succefully installed a new tracer provider.")
```
```suggestion
	level.Info(m.logger).Log("msg", "Successfully installed a new tracer provider.")
```

LeviHarrison reviewed (commented):
Thanks! Just a quick pass for now, I'll probably review more in-depth later.

matej-g commented:
Thanks for the reviews so far, I added another round of commits to improve and address suggestions (mismatch check between config and feature flag; added defaults and more clear config; documentation)

LeviHarrison commented on docs/configuration/configuration.md:
```diff
@@ -2648,6 +2648,28 @@ relabel_configs:
+# Endpoint to send the traces to. By default, this is the Otel Collector's gRPC port 7314 on a localhost.
+[ endpoint: <string> | default = localhost:7314 ]
+
+# Sets the probability a given trace will be sampled. Must be a float between 0 and 1. The default is 0, meaning no traces are exported.
```
The range is inclusive of 0 and 1 https://pkg.go.dev/go.opentelemetry.io/otel/sdk@v1.2.0/trace#TraceIDRatioBased.

```suggestion
# Sets the probability a given trace will be sampled. Must be a float from 0 through 1. The default is 0, meaning no traces are exported.
```

LeviHarrison replied:
As pointed out by @roidelapluie, having a sampling fraction of 0 means Prometheus won't create any new traces but still continue/export existing traces from other services.

```suggestion
# Sets the probability a given trace will be sampled. Must be a float between 0 and 1.
```

matej-g replied:
Agreed, that's a more precise statement on behavior, I'll remove the second sentence.

LeviHarrison reviewed (commented):
While I agree with the decision to make tracing a "manager" component, I think it needs to be further structured to behave more like one.

Manager components usually have a blocking [`Run()` method](https://github.com/prometheus/prometheus/blob/d1eadf10850ffefae0ba40421216affcda50f2e9/rules/manager.go#L930-L934), and a sometimes a [`Stop()` method](https://github.com/prometheus/prometheus/blob/d1eadf10850ffefae0ba40421216affcda50f2e9/rules/manager.go#L940-L956), which is actually already here albeit under a different name. These two are then specified in the list of routines to start: https://github.com/prometheus/prometheus/blob/main/cmd/prometheus/main.go#L766-L1066.

While the tracing manager doesn't really have any background processes that need to be started before configuration is introduced, neither (as far as I can see) does the rule manager, yet it still has a `Run()` method.

cc @roidelapluie for your opinion

matej-g commented:
Hey @LeviHarrison,
> While I agree with the decision to make tracing a "manager" component, I think it needs to be further structured to behave more like one.

If you think this makes sense from the standpoint of code comprehensibility and keeping the manager behaviors identical, I'd be happy to adjust the methods with a blocking `Run()`. However, precisely because there is no long running operation here - the "manager" only takes care of (re)installing the tracer provider, which is a one-time thing. I thought it would be redundant to have a blocking method to "run", when in the end it would just be waiting for a signal to shut down (which is instead handled with a simple `defer shutdownFunc()`). Maybe calling it "manager" is also adding to the confusion, because it's not "managing" things in the same manner as the others :slightly_smiling_face:.

matej-g commented on config/config.go:
```diff
@@ -77,6 +77,24 @@ func Load(s string, expandExternalLabels bool, logger log.Logger) (*Config, erro
 		return nil, err
 	}
 
+	// If tracing flag is not set, but the configuration is provided, return an error.
+	// Otherwise, if the flag is set, use defaults if a parameter is missing.
+	if !tracingEnabled {
+		if !cfg.TracingConfig.isZero() {
+			return nil, errors.New("tracing is configured, but the feature flag for tracing has not been set")
+		}
+	} else {
+		if cfg.TracingConfig.ServiceName == "" {
+			cfg.TracingConfig.ServiceName = DefaultTracingConfig.ServiceName
+		}
+		if cfg.TracingConfig.Endpoint == "" {
+			cfg.TracingConfig.Endpoint = DefaultTracingConfig.Endpoint
+		}
+		if cfg.TracingConfig.ClientType == "" {
+			cfg.TracingConfig.ClientType = DefaultTracingConfig.ClientType
+		}
+	}
```
The problem I was facing is that in the unmarshalling method, I'm not able to check if the tracing flag is set. Then if I set the defaults in the unmarshalling method, I wouldn't be able to tell if user configured tracing or we provided the defaults during unmarshalling.

LeviHarrison replied:
Following the precedent of other configuration options like `StorageConfig.ExemplarConfig`, we don't have to fail when configuration is provided for a feature flag not present.

matej-g replied:
That sounds like a good solution to me, i.e. I'll remove the flag, tracing will be opt-in based on the provided configuration, there will be note on stability in the documentation.

matej-g replied:
This has now been addressed.

LeviHarrison commented:
> However, precisely because there is no long running operation here - the "manager" only takes care of (re)installing the tracer provider, which is a one-time thing. I thought it would be redundant to have a blocking method to "run", when in the end it would just be waiting for a signal to shut down (which is instead handled with a simple `defer shutdownFunc()`).

Although that is a valid point, I think for the sake of consistency we should make the structure the same as other manager components :)

roidelapluie commented on config/config.go:
```diff
@@ -212,6 +230,12 @@ var (
+
+	DefaultTracingConfig = TracingConfig{
+		ServiceName: "prometheus",
+		Endpoint:    "localhost:4317",
```
The default should be empty, which means that tracing is completely disabled.

matej-g replied:
Now that the tracing will be opt-in based on whether config is provided, I agree. I'll remove this.

matej-g replied:
Done, defaults removed.

matej-g commented on config/config.go:
```diff
@@ -212,6 +230,12 @@ var (
 	}
+
+	DefaultTracingConfig = TracingConfig{
+		ServiceName: "prometheus",
```
Just speaking from personal experience, yes, users mostly seem to use the default name but sometimes it's useful to have option to override this. Nevertheless, we can always add it later if anyone requests so, so I agree with simplifying and keeping this unconfigurable for now.

matej-g replied:
Service name option has been removed.

LeviHarrison commented on tracing/tracing.go:
```diff
@@ -0,0 +1,187 @@
+}
+
+// Stop gracefully shuts down the tracer provider and stops the tracing manager.
+func (m *Manager) Stop() {
+	defer close(m.done)
+
+	if m.shutdownFunc == nil {
+		return
+	}
+
+	if err := m.shutdownFunc(); err != nil {
+		level.Error(m.logger).Log("msg", "failed to shut down the tracer provider", "err", err)
+	}
+
+	level.Info(m.logger).Log("msg", "Tracing manager stopped")
+}
```
It doesn't really matter because there aren't any background processes anyway, but it is technically inaccurate to declare that the manager has been shut down before the channel has been closed.

```suggestion
func (m *Manager) Stop() {
	defer func() {
		close(m.done)
		level.Info(m.logger).Log("msg", "Tracing manager stopped")
	}()

	if m.shutdownFunc == nil {
		return
	}

	if err := m.shutdownFunc(); err != nil {
		level.Error(m.logger).Log("msg", "failed to shut down the tracer provider", "err", err)
	}
}
```

matej-g commented on promql/engine.go:
```diff
@@ -1164,8 +1163,9 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
-	span, _ := opentracing.StartSpanFromContext(ev.ctx, stats.InnerEvalTime.SpanOperation()+" eval "+reflect.TypeOf(expr).String())
-	defer span.Finish()
+	ctxWithSpan, span := otel.Tracer("").Start(ev.ctx, stats.InnerEvalTime.SpanOperation()+" eval "+reflect.TypeOf(expr).String())
+	ev.ctx = ctxWithSpan
```
I believe the practical difference here is how we are recording the relationship between the internal evaluation spans. Since (depending on the selected `case` based on the expression type) the `eval` function might be called from within itself, we might be starting multiple spans as part of the inner evaluation. Without setting `ev.ctx`, these are recorded as siblings (see the _before_ image for illustration), which is the 'side-effect' of not passing the new context with span. On the contrary, if the `ev.ctx` is updated, the spans will have parent-child relation (see the _after_ image). Since these inner evaluation operations themselves are actually spawning more inner evaluation operations, my take is that parent-child representation is more accurate.

Before
![before-spans](https://user-images.githubusercontent.com/38492574/143787126-c271389d-2500-40c3-b2e1-510c3e4b9fb6.png)

After
![spans-after](https://user-images.githubusercontent.com/38492574/143787133-074d0529-6146-47c5-abd8-a9fc827cba5e.png)

roidelapluie replied:
if you do 1+1, they should probably not be parent-child.

matej-g replied:
Could you please elaborate on what you mean by the 1+1 example?

matej-g commented:
gentle ping @roidelapluie @LeviHarrison, please let me know if there are any more concerns regarding this PR (apart from the unresolved parent-child spans question above, but I'm happy to adjust that) so that we can move it forward. Thanks!

LeviHarrison reviewed (commented):
Sorry about the delay, I think this is getting really close to being done.

matej-g commented on tracing/tracing.go:
```diff
@@ -0,0 +1,187 @@
+// getClient returns an appropriate OTLP client (either gRPC or HTTP), based
+// on the provided tracing configuration.
+func getClient(tracingCfg config.TracingConfig) otlptrace.Client {
+	var client otlptrace.Client
+	switch tracingCfg.ClientType {
+	case config.TracingClientGRPC:
+		opts := []otlptracegrpc.Option{otlptracegrpc.WithEndpoint(tracingCfg.Endpoint)}
+		if !tracingCfg.WithSecure {
+			opts = append(opts, otlptracegrpc.WithInsecure())
+		}
+
+		client = otlptracegrpc.NewClient(opts...)
+	case config.TracingClientHTTP:
+		opts := []otlptracehttp.Option{otlptracehttp.WithEndpoint(tracingCfg.Endpoint)}
+		if !tracingCfg.WithSecure {
+			opts = append(opts, otlptracehttp.WithInsecure())
+		}
+		client = otlptracehttp.NewClient(opts...)
+	}
+
+	return client
```
There's a difference between each case, since once we're calling `NewClient` for gRPC (from `otlptracegrpc` package), the other time it's called from `otlptracehttp` for HTTP client. 

With this change, we would always be creating HTTP client from `otlptracehttp.NewClient`.

LeviHarrison replied:
Ah didn't notice that, thanks.

matej-g commented on tracing/tracing_test.go:
```diff
@@ -0,0 +1,66 @@
+		},
+	}
+
+	m.ApplyConfig(&cfg)
```
Thanks, adjusted.

matej-g commented on tracing/tracing_test.go:
```diff
@@ -0,0 +1,66 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracing
```
Good point, I added test cases for both (uninstall, shut down).

matej-g commented on docs/configuration/configuration.md:
````diff
@@ -2648,6 +2648,23 @@ relabel_configs:
   [ - <relabel_config> ... ]
 ```
 
+### `<tracing>`
````
Thanks for pointing this out :+1: I adjusted to documentation accordingly.

LeviHarrison commented on docs/configuration/configuration.md:
```diff
@@ -98,7 +98,23 @@ remote_read:
-  [ - <exemplars> ... ]  
+  [ - <exemplars> ... ]
+
+# Tracing configures exporting traces from Prometheus to a tracing
+# backend via the OTLP protocol. Tracing is currently an experimental
+# feature and could change in the future.
+tracing:
+  # Client used to export the traces. Options are 'http' or 'grpc'.
+  client_type: <string>
+
+  # Endpoint to send the traces to.
+  endpoint: <string>
+
+  # Sets the probability a given trace will be sampled. Must be a float from 0 through 1.
+  [ sampling_fraction: <float> | default = 0 ]
+
+  # If enabled, the client will use a secure connection.
+  [ with_secure: <bool> | default = false ]
```
Sorry, my request was unclear here. What I meant was a stub like this:

```yaml
tracing:
  [ <tracing> ]
```

I think the config should still have a separate section like before.

matej-g replied:
Fixed, thanks :raised_hands:

LeviHarrison reviewed (approved):
Good from me after this small config nitpick. I just tested and it works great.

LeviHarrison commented:
If anyone is looking for a quick way to test this out, build Prometheus from this branch, then add this to the example config file:

```diff
diff --git a/documentation/examples/prometheus.yml b/documentation/examples/prometheus.yml
index 312b578aa..05c2a9ffb 100644
--- a/documentation/examples/prometheus.yml
+++ b/documentation/examples/prometheus.yml
@@ -27,3 +27,8 @@ scrape_configs:
 
     static_configs:
       - targets: ["localhost:9090"]
+
+tracing:
+  client_type: "http"
+  endpoint: "localhost:55681"
+  sampling_fraction: 1
```

Start a Jaeger OLTP instance:

```
docker run --name jaeger \
  -p 16686:16686 \
  -p 55681:55681 \
  jaegertracing/opentelemetry-all-in-one
```

Run Prometheus:

```
./prometheus --config.file=./documentation/examples/prometheus.yml
```

And view traces in the Jaeger web UI: http://localhost:16686

LeviHarrison commented:
Oh and also, does [this file](https://github.com/prometheus/prometheus/blob/f041a6a228b1a7e34e1337e5c0974db63820c6cb/util/stats/query_stats.go#L19) need to be converted to the OpenTelemetry library?

matej-g commented:
> Oh and also, does [this file](https://github.com/prometheus/prometheus/blob/f041a6a228b1a7e34e1337e5c0974db63820c6cb/util/stats/query_stats.go#L19) need to be changed to use the OpenTelemetry library?

Unless I'm missing something, this has been already updated in my branch (https://github.com/matej-g/prometheus/blob/migrate-tracing-to-otel/util/stats/query_stats.go#L20) :thinking:. There should be no more files using the OpenTracing instrumentation anywhere.

LeviHarrison commented:
> Unless I'm missing something, this has been already updated in my branch (https://github.com/matej-g/prometheus/blob/migrate-tracing-to-otel/util/stats/query_stats.go#L20) 🤔

Yup, sorry, didn't see that in my changes tab for some reason.

matej-g commented on docs/configuration/configuration.md:
```diff
@@ -98,7 +98,13 @@ remote_read:
+
+# Tracing configures exporting traces from Prometheus to a tracing
+# backend via the OTLP protocol. Tracing is currently an experimental
+# feature and could change in the future.
```
Amended, thanks!

roidelapluie commented:
Thanks. I see windows tests failing, could you rebase? I added more comments, and I am testing it now.

roidelapluie commented on config/config.go:
```diff
@@ -499,6 +500,57 @@ type StorageConfig struct {
+type TracingConfig struct {
+	ClientType       TracingClientType `yaml:"client_type,omitempty"`
+	Endpoint         string            `yaml:"endpoint,omitempty"`
+	SamplingFraction float64           `yaml:"sampling_fraction,omitempty"`
```
Is there any other sampling methods supported?

matej-g commented on config/config.go:
```diff
@@ -499,6 +500,57 @@ type StorageConfig struct {
+	}
+
+	if t.ClientType == "" {
+		return errors.New("tracing client type must be set")
```
Made gRPC the default, thanks.

matej-g commented on config/config.go:
```diff
@@ -499,6 +500,57 @@ type StorageConfig struct {
+	ClientType       TracingClientType `yaml:"client_type,omitempty"`
+	Endpoint         string            `yaml:"endpoint,omitempty"`
+	SamplingFraction float64           `yaml:"sampling_fraction,omitempty"`
+	WithSecure       bool              `yaml:"with_secure,omitempty"`
```
Added TLS config as well.

roidelapluie commented on tracing/tracing.go:
```diff
@@ -0,0 +1,205 @@
+// the current tracer provider (if any is registered) and installing a new one.
+func (m *Manager) ApplyConfig(cfg *config.Config) error {
+	// Update only if a config change is detected.
+	if m.config == cfg.TracingConfig {
```
I think with a TLS config this is not true because of some private functions in the TLS configs. Is this tested?

LeviHarrison replied:
`TestReinstallingTracerProvider` should handle it: https://github.com/matej-g/prometheus/blob/54ec999c98f9abb7d71939d2052dcdd6e7029652/tracing/tracing_test.go#L42-L67

(Although maybe including a TLS config will change something?)

matej-g replied:
Hm, all the `config.TLSConfig` struct fields are exported and can be compared (since its `string` and `bool` fields), so I would expect this comparison to work fine. I added a TLS config to the reinstall test case, which as @LeviHarrison should caught this in case the comparison would not work as expected.

roidelapluie commented:
Thanks! Let's take it from here.

vsxen commented:
OpenTelemetry looks not work with grpc and insecure.
```
# prom 
tracing:
  endpoint: localhost:4317
  client_type: grpc
  insecure: true
  sampling_fraction: 1

```
prom log
ts=2022-04-06T07:19:20.113Z caller=tracing.go:64 level=error msg="OpenTelemetry handler returned an error" err="max retry time elapsed: rpc error: code = Unavailable desc = connection error: desc = \"transport: authentication handshake failed: tls: first record does not look like a TLS handshake\""

```
receivers:
  otlp:
    protocols:
      grpc:
      http:

exporters:
  jaeger:
    endpoint: "localhost:14250"
    tls:
      insecure: true

  logging:
    loglevel: debug

service:
  pipelines:
    traces/1:
      receivers: [otlp]
      exporters: [jaeger]
```
otel log
2022-04-06T15:17:35.078+0800	warn	zapgrpc/zapgrpc.go:191	[core] grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: http2Server.HandleStreams received bogus greeting from client: \"\\x16\\x03\\x01\\x01\\t\\x01\\x00\\x01\\x05\\x03\\x03P\\xc7ǅSV\\x95)\\x15HXe\\x96\""	{"grpc_log": true}
2022-04-06T15:18:05.082+0800	warn	zapgrpc/zapgrpc.go:191	[core] grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: http2Server.HandleStreams received bogus greeting from client: \"\\x16\\x03\\x01\\x01\\t\\x01\\x00\\x01\\x05\\x03\\x03\\xe3\\x1a\\x04\\x15Lݤ\\xb8\\xd0\\nw)\\xf0\""	{"grpc_log": true}

roidelapluie commented:
Can you please open a dedicated issue for this? Thanks.

Le mer. 6 avr. 2022, 09:46, vxsen robot ***@***.***> a écrit :

> OpenTelemetry looks not work with grpc and insecure.
>
> # prom
>
> tracing:
>
>   endpoint: localhost:4317
>
>   client_type: grpc
>
>   insecure: true
>
>   sampling_fraction: 1
>
>
>
>
> prom log
> ts=2022-04-06T07:19:20.113Z caller=tracing.go:64 level=error
> msg="OpenTelemetry handler returned an error" err="max retry time elapsed:
> rpc error: code = Unavailable desc = connection error: desc = "transport:
> authentication handshake failed: tls: first record does not look like a TLS
> handshake""
>
> receivers:
>
>   otlp:
>
>     protocols:
>
>       grpc:
>
>       http:
>
>
>
> exporters:
>
>   jaeger:
>
>     endpoint: "localhost:14250"
>
>     tls:
>
>       insecure: true
>
>
>
>   logging:
>
>     loglevel: debug
>
>
>
> service:
>
>   pipelines:
>
>     traces/1:
>
>       receivers: [otlp]
>
>       exporters: [jaeger]
>
>
> otel log
> 2022-04-06T15:17:35.078+0800 warn zapgrpc/zapgrpc.go:191 [core] grpc:
> Server.Serve failed to create ServerTransport: connection error: desc =
> "transport: http2Server.HandleStreams received bogus greeting from client:
> "\x16\x03\x01\x01\t\x01\x00\x01\x05\x03\x03P\xc7ǅSV\x95)\x15HXe\x96""
> {"grpc_log": true}
> 2022-04-06T15:18:05.082+0800 warn zapgrpc/zapgrpc.go:191 [core] grpc:
> Server.Serve failed to create ServerTransport: connection error: desc =
> "transport: http2Server.HandleStreams received bogus greeting from client:
> "\x16\x03\x01\x01\t\x01\x00\x01\x05\x03\x03\xe3\x1a\x04\x15Lݤ\xb8\xd0\nw)\xf0""
> {"grpc_log": true}
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/prometheus/pull/9724#issuecomment-1089936744>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACHHJUZ2H7SSL3YFRTLYZLVDU6MRANCNFSM5H5BWMHQ>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

vsxen commented:
https://github.com/prometheus/prometheus/issues/10542
